funfact/lang/interpreter/_shape.py

```diff
@@ -50,12 +50,35 @@ def pow(self, base: P.Numeric, exponent: P.Numeric, **kwargs):
     def neg(self, x: P.Numeric, **kwargs):
         return x.shape
 
+    @as_payload
+    def matmul(self, lhs: P.Numeric, rhs: P.Numeric, **kwargs):
+        if len(lhs.shape) != 2:
+            raise SyntaxError(
+                'Only matrices (2D tensors) supports multiplication '
+                f'using `@`. Left-hand side operand has shape {lhs.shape}.'
+            )
+        if len(rhs.shape) != 2:
+            raise SyntaxError(
+                'Only matrices (2D tensors) supports multiplication '
+                f'using `@`. Right-hand side operand has shape {rhs.shape}.'
+            )
+        if lhs.shape[1] != rhs.shape[0]:
+            raise SyntaxError(
+                f'Dimensions of matrices {lhs.shape} and {rhs.shape} '
+                'not compatible for multiplication.'
```

Comment:
maybe add `using '@'` to this error message as well?

Reply:
Done.